compiler/transforms/reuse_memref_space.py

```diff
@@ -0,0 +1,254 @@
+from collections.abc import Sequence
+
+from xdsl.context import MLContext
+from xdsl.dialects import affine, arith, builtin, memref, scf
+from xdsl.dialects.builtin import IndexType
+from xdsl.ir import Block, Operation, OpResult, SSAValue
+from xdsl.ir.affine import AffineConstantExpr
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.rewriter import InsertPoint
+
+
+class MoveMemrefAllocations(RewritePattern):
+    """
+    This class represents a rewrite pattern for moving memref allocations outside
+    a double for-loop. This is possible when each loop a new memref-space is allocated with
+    identical sizes and the space is not used outside the loop.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, alloc_op: memref.Alloc, rewriter: PatternRewriter):
+        def find_parent_for_loop(op: Operation) -> scf.For:
+            """
+            Find the parent for-loop of the operation.
+            If no for-loop is found, return None.
+            """
+            op = op.parent_op()
+            while not isinstance(op, scf.For):
+                if op.parent_op() is None:
+                    return None
+                op = op.parent_op()
+            return op
```

Comment:
These are all nice helper functions, but define them outside of the class if they do not depend on it!